src/lib/pure_pursuit/PurePursuit.cpp

```diff
@@ -0,0 +1,112 @@
+ *   Copyright (c) 2024 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "PurePursuit.hpp"
+#include <mathlib/mathlib.h>
+
+float PurePursuit::calcDesiredHeading(const Vector2f &curr_wp_ned, const Vector2f &prev_wp_ned,
+				      const Vector2f &curr_pos_ned,
+				      const float &lookahead_distance)
+{
+	// Check minimal necessary input validity
+	if (!curr_wp_ned.isAllFinite() || !curr_pos_ned.isAllFinite()) {
+		return 0.f; // No fallback possible
+	}
+
+	// Setup fallback to bearing to the current waypoint
+	const Vector2f curr_pos_to_curr_wp = curr_wp_ned - curr_pos_ned;
+	const float bearing_to_curr_wp = atan2f(curr_pos_to_curr_wp(1), curr_pos_to_curr_wp(0));
+
+	// Check full input validity
+	if (lookahead_distance < FLT_EPSILON || !PX4_ISFINITE(lookahead_distance) || !prev_wp_ned.isAllFinite()) {
+		return bearing_to_curr_wp; // Fallback
+	}
+
+	// Catch waypoint overlap
+	if (fabsf(curr_wp_ned(0) - prev_wp_ned(0)) < FLT_EPSILON
+	    && fabsf(curr_wp_ned(1) - prev_wp_ned(1)) < FLT_EPSILON) {
+		return bearing_to_curr_wp; // Fallback
+	}
+
+	// Catch infinite line segment slope (Current and previous waypoints have same north coordinate)
+	if (curr_wp_ned(0) - prev_wp_ned(0) < FLT_EPSILON) {
+		if (fabsf(curr_wp_ned(0) - curr_pos_ned(0)) > lookahead_distance) { // No intersection point exists
+			return bearing_to_curr_wp; // Fallback
+
+		} else {
```

Comment:
and what if the distance between the current wp and position is < lookahead_distance ? Don't you want to directly go to the current wp?

Reply:
This is also an option, but has this tradeoff:
This will essentially reduce the lookahead radius until the vehicle enters the acceptance radius. If you have a small acceptance radius this can lead to oscillations (general problem with small lookahead).
Might still be the better option, I'll test it with the ackermann rover.

Reply:
I now implemented this in 29b93d9, i agree that it is the better option.